compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/GPU/TargetUtils/ConfigUtils.h"
+#include <numeric>
+
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "llvm/Support/Casting.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+namespace mlir::iree_compiler::IREE::GPU {
+
+static constexpr int64_t kPreferredCopyNumBits = 128;
+
+SmallVector<int64_t> deriveThreadTileSizes(Operation *op) {
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp || !linalgOp.hasPureTensorSemantics()) {
+    return {};
+  }
+
+  // TODO: Support multi-result
+  if (linalgOp->getNumResults() != 1) {
+    return {};
+  }
+
+  SmallVector<int64_t> loopRanges = linalgOp.getStaticLoopRanges();
+  // TODO: We shouldn't need this check, however loop fusion currently requires
+  // loop trip counts to be identical, meaning we need to use a num_threads
+  // variant of tiling. Remove this and simply return the preferred vector size
+  // once loop fusion can resolve the forall properly.
+  if (llvm::any_of(loopRanges,
+                   [](int64_t s) { return ShapedType::isDynamic(s); })) {
+    return {};
+  }
+
+  std::optional<SmallVector<int64_t>> workgroupSize =
+      getWorkgroupSize(op->getParentOfType<FunctionOpInterface>());
+  if (!workgroupSize) {
```

Comment:
What happens if the workgroup size is not static?

Reply:
It won't work right now but we would probably just either emit code similar to the below to compute the tile sizes, or just return a static tile size of `kPreferredCopyNumBits / elementTypeBitWidth` and resolve the `scf.forall` to a loop (maybe with a #pragma unroll type thing to unroll the loop if it resolves to being static later one).

For now I could add an assert here if you think that makes more sense as this is essentially unsupported (and I don't think we currently have a way to specify non-static workgroup size).

Reply:
Ah, workgroup size. I got confused with workgroup count. THis is fine then.